Add Lighthouse CL

jimmygchen opened the pull request:
Add latest Lighthouse version to `docker-compose`. 

There are some outstanding work to get this working as geth `genesis.json` & chain config are still WIP. Built on top of @realbigsean's work from this branch: https://github.com/realbigsean/eip4844-interop/tree/interop-lighthouse

Previous PR: https://github.com/Inphi/eip4844-interop/pull/47

jimmygchen committed:
add lighthouse submodule

jimmygchen committed:
add lighthouse to docker-compose

jimmygchen committed:
update geth version to latest

jimmygchen commented:
@realbigsean I've tried using the same genesis config and chain config you had on the lighthouse [local testnet](https://github.com/sigp/lighthouse/tree/eip4844/scripts/local_testnet), but had some issues with triggering the merge. I'll take another look later today.

jimmygchen committed:
generate genesis file with shanghaiTime and shanghaiForkTime

jimmygchen commented:
I created a temporary branch to test this with a generated genesis file, as `testnetgen` is yet to be merged in #74:

https://github.com/Inphi/eip4844-interop/compare/master...jimmygchen:eip4844-interop:lighthouse-interop-gen-genesis?expand=1

Still having some issues with merge, geth seems to be waiting for consensus updates, but lighthouse thinks geth isn't sync'd :/ I **think** I'm using the exact same genesis.json and chain config from Lighthouse [local testnet](https://github.com/sigp/lighthouse/tree/eip4844/scripts/local_testnet) - except for `chainId` and `depositContractAddress`

Inphi commented:
How is the lighthouse testnet generated? Is the genesis at bellatrix like how @realbigsean's PR sets up the testnet?

jimmygchen committed:
update lighthouse version and genesis generation

jimmygchen commented:
@Inphi, thanks, that was a good thing to look at, given that I'm running the same client versions & same chain config already.

I'm using the `lcli interop-genesis` command to generate the genesis, and the same [config.yaml](https://github.com/jimmygchen/eip4844-interop/commit/40fc036b8f40accd2af0b8ad32810b75c746e288#diff-42abb036bb2ad6f0586fe00927ae2da1299738a82ccd86af713717c905444257) file from @realbigsean's setup. The script can be found here:

https://github.com/jimmygchen/eip4844-interop/commit/40fc036b8f40accd2af0b8ad32810b75c746e288#diff-5a098c6b6144fcf0141f1255e8fc64cd3c36a700be26709b2e4c8c0111aa6f95

I've also tried copying the genesis.json & genesis.ssz files from @realbigsean's local testnet, but still having the same issue :/

jimmygchen committed:
fix lighthouse not transitioning through merge

jimmygchen committed:
add lighthouse client to bootstrap.go

jimmygchen commented:
@Inphi @realbigsean I've made some progress here, the beacon chain now progresses and transition through the merge!

I think the issue was I had VC pointing to the follower beacon node, which doesn't start until the 1st beacon node is ready, hence there were no blocks being proposed. Pointing the VC to the 1st beacon node fixes the problem.

The other problem I'm having now is having the 2nd execution node to sync, the 2nd geth node kept dropping the 1st geth peer. This seem to happen on other Lodestar-Geth as well, so I'm guessing it might be related to the geth upgrade in this PR

```
DEBUG[12-12|05:30:17.905] Adding p2p peer                          peercount=1 id=b54508f440d24b07 conn=dyndial addr=192.168.32.2:30303 name=Geth/v1.11.0-unstabl...
TRACE[12-12|05:30:17.905] Starting protocol eth/68                 id=b54508f440d24b07 conn=dyndial
TRACE[12-12|05:30:17.905] Starting protocol snap/1                 id=b54508f440d24b07 conn=dyndial
DEBUG[12-12|05:30:17.909] Ethereum peer connected                  id=b54508f440d24b07 conn=dyndial name=Geth/v1.11.0-unstabl...
TRACE[12-12|05:30:17.909] Registering sync peer                    peer=b54508f4
INFO [12-12|05:30:18.479] Unlocked account                         address=0xa94f5374Fce5edBC8E2a8697C15331677e6EbF0B
TRACE[12-12|05:30:18.907] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:18.909] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:19.408] FINDNODE failed                          id=b54508f440d24b07 failcount=5 dropped=false err="RPC timeout"
TRACE[12-12|05:30:19.409] FINDNODE failed                          id=b54508f440d24b07 failcount=5 dropped=false err="RPC timeout"
TRACE[12-12|05:30:19.410] >> PING/v4                               id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:19.410] >> PING/v4                               id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
DEBUG[12-12|05:30:20.004] Couldn't add port mapping                proto=tcp extport=30303 intport=30303 interface="UPnP or NAT-PMP" err="no UPnP or NAT-PMP router discovered"
DEBUG[12-12|05:30:20.005] Couldn't add port mapping                proto=udp extport=30303 intport=30303 interface="UPnP or NAT-PMP" err="no UPnP or NAT-PMP router discovered"
TRACE[12-12|05:30:20.417] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:20.418] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
DEBUG[12-12|05:30:20.828] FS scan times                            list="406.417µs" set="55.083µs" diff="8.375µs"
TRACE[12-12|05:30:20.917] FINDNODE failed                          id=b54508f440d24b07 failcount=6 dropped=false err="RPC timeout"
TRACE[12-12|05:30:20.917] FINDNODE failed                          id=b54508f440d24b07 failcount=6 dropped=false err="RPC timeout"
TRACE[12-12|05:30:20.917] >> PING/v4                               id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:21.924] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:22.423] FINDNODE failed                          id=b54508f440d24b07 failcount=7 dropped=false err="RPC timeout"
TRACE[12-12|05:30:22.424] >> PING/v4                               id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
TRACE[12-12|05:30:23.427] >> FINDNODE/v4                           id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
DEBUG[12-12|05:30:23.833] FS scan times                            list="217.334µs" set="42.791µs" diff="20.625µs"
TRACE[12-12|05:30:23.931] FINDNODE failed                          id=b54508f440d24b07 failcount=8 dropped=false err="RPC timeout"
DEBUG[12-12|05:30:26.836] FS scan times                            list="383.416µs" set="94.5µs"   diff="11.875µs"
TRACE[12-12|05:30:27.174] >> PING/v4                               id=b54508f440d24b07 addr=192.168.32.2:30303 err=nil
DEBUG[12-12|05:30:27.675] Removed dead node                        b=12 id=b54508f440d24b07 ip=192.168.32.2 checks=0
```

jimmygchen committed:
revert non lighthouse related changes

jimmygchen committed:
add lighthouse tests to CI

jimmygchen commented:
@Inphi this PR is now ready for review. I've removed the non lighthouse related changes (e.g. geth version updates) from this PR, so we can at least run the pre 4844 tests.

jimmygchen committed:
fix build error

jimmygchen committed:
increase build timeout

jimmygchen committed:
update trusted setup

jimmygchen committed:
revert chain config change

jimmygchen commented:
At the moment the beacon node continues to produce block until Capella, and stopped on this error: `method engine_forkchoiceUpdatedV2 does not exist/is not available`, so looks like we'll need the new geth version that supports withdrawal.

roberto-bayardo commented:
> At the moment the beacon node continues to produce block until Capella, and stopped on this error: `method engine_forkchoiceUpdatedV2 does not exist/is not available`, so looks like we'll need the new geth version that supports withdrawal.

Can you try interoperating with the head of the eip-4844 branch of our fork?

realbigsean commented:
I think it's good adding the trusted setup file to the shared directory, but would be better to use the format in the spec repo here, because I think other clients are probably using this format: https://raw.githubusercontent.com/ethereum/consensus-specs/dev/presets/mainnet/trusted_setups/testing_trusted_setups.json

Lighthouse can't actually read the JSON format at the moment though, pawan is working on making it so that we can. But until then we can point `--trusted-setup-file` to `lighthouse/scripts/local_testnet/trusted_setup.txt`

jimmygchen committed:
Merge branch 'lighthouse-interop' into lighthouse-interop-gen-genesis

jimmygchen committed:
Update geth to latest eip-4844 branch

jimmygchen committed:
fix missing genesis file

jimmygchen commented:
> Can you try interoperating with the head of the eip-4844 branch of our fork?

I'm now getting these errors with the [head of eip-4844 branch](https://github.com/mdehoog/go-ethereum/commit/b697b4457653088319be37bec5eb85250036cb58)

```
WARN [12-13|02:26:08.505] Served engine_exchangeTransitionConfigurationV1 conn=172.23.0.5:41228 reqid=1 duration="25.917µs"  err="the method engine_exchangeTransitionConfigurationV1 does not exist/is not available"
...
WARN [12-13|02:26:09.012] Served engine_forkchoiceUpdatedV1        conn=172.23.0.5:41228 reqid=1 duration="23.209µs"  err="the method engine_forkchoiceUpdatedV1 does not exist/is not available"
```

looks like this is related to the latest [merged MR](https://github.com/mdehoog/go-ethereum/pull/79/files#diff-35925471aabda6d82266ade49d1604a5be602ff33c0ef8913e443ad1942a41d6) re-enabling authentication. I'll try adding auth params and see if it works

jimmygchen committed:
more fixes

jimmygchen commented:
So I've made a few more changes to the experimental branch here:
https://github.com/jimmygchen/eip4844-interop/pull/1

This branch contains the following additional changes
1. Generate geth genesis with computed  `shanghaiTime` and `shanghaiForkTime`
2. Updated Lighthouse & Geth to head of eip-4844 branches
3. Added JWT authentication (required now for geth)

The engine API errors in the above comment are now gone, however I can't get the chain to advance and I'm a bit stuck here. I'm seeing errors all over the place, but it seems like it may be related to the geth nodes not syncing with each other -  the 2nd peer node doesn't seem to be able to get a PING response from the 1st geth node, and drops the node after a few attempts.

Inphi commented:
> > At the moment the beacon node continues to produce block until Capella, and stopped on this error: `method engine_forkchoiceUpdatedV2 does not exist/is not available`, so looks like we'll need the new geth version that supports withdrawal.
> 
> Can you try interoperating with the head of the eip-4844 branch of our fork?

Tested it using the latest eip-4844 geth. While the beacon chain upgrades to EIP4844, I'm seeing the following errors:
```
eip4844-interop-lighthouse-beacon-node-1 | Dec 12 23:54:59.007 INFO Prepared beacon proposer                validator: 1, prepare_slot: 129, already_known: false, service: beacon
eip4844-interop-lighthouse-beacon-node-1 | Dec 12 23:54:59.035 ERRO No valid eth1_data votes, `votes_to_consider` empty, outcome: casting `state.eth1_data` as eth1 vote, genesis_time: 1670887763, earliest_block_timestamp: 0, lowest_block_number: 0, service: deposit_contract_rpc
eip4844-interop-lighthouse-beacon-node-1 | Dec 12 23:54:59.036 ERRO Execution engine call failed            error: BadConversion("Null `excess_data_gas` field converting JsonExecutionPayloadV2 -> ExecutionPayloadEip4844"), service: exec
```

Though this could be because the `shardingForkTime` isn't configured correctly?

Inphi committed:
Working Lighthouse devnet

Inphi committed:
update geth/lighthouse dependencies

jimmygchen commented:
Hey @Inphi, thanks for looking into this! Are you using the same branch in this PR? https://github.com/jimmygchen/eip4844-interop/pull/1

`shardingForkTime` is computed [in this script](https://github.com/jimmygchen/eip4844-interop/blob/f665a5edca7f8b39b94540ff26b3f3c2c85b0974/lighthouse/generate-genesis.sh#L26) - same as what Sean is using for Lighthouse local testnet, and I'm seeing values being populated on Geth startup.

I'm going to continue investigation on this - is there any ETA on when #74 would be ready, so I can test with the exact same setup that works for other clients?

Inphi commented:
hey @jimmygchen. I got the a lighthouse/geth devnet working via https://github.com/jimmygchen/eip4844-interop/pull/2.
The `lcli-build-genesis` service can generate the configs we need for client interop. So I'd like to use that instead of #74.

I'm able to trigger the EIP-4844 upgrade, send blob transactions and confirm their commitments on the beacon chain. The last component that we need to run all tests is to be able to download blobs sidecars from the beacon chain. I'm running into issues doing this using the p2p Req-Res interface. Does lighthouse expose a HTTP beacon chain API to retrieve blobs? I could use that instead of Req-Res.

jimmygchen commented:
wow nice work!

hmm I wonder why downloading via p2p Req-Res interface doesn't work, I believe it worked for Prysm in earlier testnets? what issue are you seeing? FYI: @realbigsean 

I don't think Lighthouse expose this Beacon API endpoint yet - in fact, I've just created an issue for this in the beacon API spec earlier today:
https://github.com/ethereum/beacon-APIs/issues/282

Inphi commented:
I'm getting a "protocol not supported" error when attempting to create a stream for a `blobs_sidecars_by_range` request. According to the logs, the lighthouse beacon nodes do not support the `blobs_sidecars_by_range` Req-Res, or at least they aren't advertising it:

```
Dec 16 07:15:30.003 DEBG Identified Peer                         protocols: ["/meshsub/1.1.0", "/meshsub/1.0.0", "/eth2/beacon_chain/req/status/1/ssz_snappy", "/eth2/beacon_chain/req/goodbye/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/1/ssz_snappy", "/eth2/beacon_chain/req/ping/1/ssz_snappy", "/eth2/beacon_chain/req/metadata/2/ssz_snappy", "/eth2/beacon_chain/req/metadata/1/ssz_snappy", "/ipfs/id/1.0.0", "/ipfs/id/push/1.0.0"], observed_address: "/ip4/127.0.0.1/tcp/9001/p2p/16Uiu2HAmCHMi7ZK5T4pLepymn7fp3W4bydedBCS5Tva131EAS6AQ", listening_addresses: ["/ip4/127.0.0.1/tcp/9002", "/ip4/172.24.53.62/tcp/9002", "/ip4/127.0.0.1/tcp/9002", "/ip4/172.17.0.1/tcp/9002"], agent_version: Lighthouse/v3.3.0-1644978+/x86_64-linux, protocol_version: eth2/1.0.0, peer: 16Uiu2HAkzNWuYNSoJLBxrpT1eBRsyMRJWjhW1AworbybRqSN94Nd, service: libp2p
```

`/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy` is on that list, but not `/eth2/beacon_chain/req/blobs_sidecars_by_range/1/ssz_snappy` (and neither is `/eth2/beacon_chain/req/beacon_block_and_blobs_sidecar_by_root/1` which is needed for historical sync)

Inphi commented:
This is probably a bug in lighthouse. I setup a lighthouse beacon node to trigger historical sync so it also issues `blobs_sidecar_by_root` requests. The beacon node stalls with this error:
```
Dec 16 16:59:15.212 DEBG Sending BlockBlock by range request     peer: 16Uiu2HAm5rqrdxgrwSRfeX3wNJ42sZR95kh5sdzX91GgY7qwSzar, count: 32, method: Mixed by range request, service: sync
Dec 16 16:59:15.212 DEBG Requesting batch                        start_slot: 64, end_slot: 95, downloaded: 3, processed: 0, processed_no_penalty: 0, state: Downloading(16Uiu2HAm5rqrdxgrwSRfeX3wNJ42sZR95kh5sdzX91GgY7qwSzar, 0 blocks, 6), epoch: 2, chain: 7817711998956517890, service: sync
Dec 16 16:59:15.215 DEBG RPC Error                               direction: Outgoing, score: 0, peer_id: 16Uiu2HAm5rqrdxgrwSRfeX3wNJ42sZR95kh5sdzX91GgY7qwSzar, client: Unknown, err: Peer does not support the protocol, protocol: blobs_sidecars_by_range, service: libp2p
Dec 16 16:59:15.215 DEBG Batch failed. RPC Error                 batch_epoch: 2, chain: 7817711998956517890, service: sync
```

realbigsean commented:
> The `lcli-build-genesis` service can generate the configs we need for client interop. So I'd like to use that instead of #74.

Awesome! I've actually been playing around with this tool that EF devops is using for capella testnets. I think this might be preferable to the `lcli` thing we have going. Notably it's able to produce validator keystores in the correct format for each CL client which is very useful for actually distributing the keys to different VCs 
https://github.com/ethpandaops/ethereum-genesis-generator

jimmygchen committed:
Update Lighthouse version

kanaagawa commented on geth/geth-genesis.json:
```diff
@@ -3,42 +3,853 @@
     "chainId": 1,
```
Why are we using a different `chainId` here and in `./lighthouse/genesis.json`?

kanaagawa replied:
After doing a more involved review I realize we are using the `generated-genesis.json` file now. I wonder if we should remove this to prevent confusion? Or we think is still relevant in other use cases?

jimmygchen replied:
it's not used by lighthouse, but used by other clients - there are parallel work happening in #78.. I think eventually we'll use the same one

kanaagawa commented on .github/workflows/ci.yml:
```diff
@@ -113,3 +113,56 @@ jobs:
+        timeout-minutes: 60
+        run: go run ./tests/pre-4844 lighthouse
+
+      # - name: Lighthouse - Run Blob transaction tests
```
Maybe is a good time to consider running this tests?

jimmygchen replied:
👍

kanaagawa commented on docker-compose.yml:
```diff
@@ -5,38 +5,66 @@ volumes:
+    entrypoint: ["geth.sh"]
     volumes:
       - 'geth_data2:/db'
+      - 'genesis_data:/genesis_data'
```
```suggestion
```

kanaagawa replied:
I think we can remove this with the new setup.

kanaagawa commented on docker-compose.yml:
```diff
@@ -5,38 +5,66 @@ volumes:
   geth_data2:
   beacon_node_data:
   beacon_node_follower_data:
+  genesis_data:
```
```suggestion
```

kanaagawa commented on geth/geth.sh:
```diff
@@ -2,77 +2,48 @@
-BOOTNODE_KEY_HEX=${BOOTNODE_KEY_HEX:-65f77f40c167b52b5cc70fb33582aecbdcd81062dc1438df00a3099a07079204}
-NETWORKID=69
-ENABLE_MINING=true
+GETH_GENESIS=/shared-configs/genesis.json
```
Maybe I'm not following the setup correctly, but shouldn't this be `generated-genesis.json` instead?

Inphi replied:
It could go either way. This works now because the `generated-genesis.json` is a copy of `/shared-configs/genesis.json`. Actually `/shared-configs/genesis.json` is the master file which is located on the lcli testnet directory.

One issue I'm noticing with docker is that it take a couple seconds to synchronize bind mounts. And `generated-genesis.json` is a bind-mount to the lcli testnet directory. So using the testnet directory here rather than `generated-genesis.json` prevents a race condition for the mount sync.

kanaagawa commented on lighthouse/genesis.json:
```diff
@@ -0,0 +1,858 @@
+    },
+    "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b": {
+      "balance": "0x6d6172697573766477000000"
+    }
```
```suggestion
    }
```

kanaagawa replied:
It was repeated.

kanaagawa commented on lighthouse/generate-genesis.sh:
```diff
@@ -0,0 +1,73 @@
+	--count $VALIDATOR_COUNT \
+	--base-dir $DATADIR \
+    --testnet-dir $TESTNET_DIR \
+	--node-count 1
```
```suggestion
	--count $VALIDATOR_COUNT \
	--base-dir $DATADIR \
	--testnet-dir $TESTNET_DIR \
	--node-count 1
```

kanaagawa commented on docker-compose.yml:
```diff
@@ -183,6 +211,125 @@ services:
+      - lighthouse-bootnode
+      - execution-node
+      - lcli-build-genesis
+    # build:
+    #   context: ./lighthouse
+    #   dockerfile: Dockerfile.lighthouse
```
What do you think of remove this section now?

kanaagawa commented on docker-compose.yml:
```diff
@@ -183,6 +211,125 @@ services:
+    depends_on:
+      - execution-node-2
+      - lighthouse-beacon-node
+    # build:
+    #   context: ./lighthouse
+    #   dockerfile: Dockerfile.lighthouse
```
Same as above.

kanaagawa commented on docker-compose.yml:
```diff
@@ -5,38 +5,66 @@ volumes:
+ 
   execution-node:
+    depends_on:
+      #- execution-bootnode
```
I think we still need the `execution-node` to depend on the `execution-bootnode` right?

Inphi replied:
I couldn't get geth to connect to the bootnode.
The original reason to have both EL clients peer was to synchronize blocks pre-merge. But now that we're post-merge at genesis, we no longer really need to do this.

It'll be nice to have the bootnode working though, so we can test the 4844 mempool across multiple EL implementations. But for now this is OK.

kanaagawa commented on geth/bootnode.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+priv_key="02fd74636e96a8ffac8e7b01b0de8dea94d6bcf4989513b38cf59eb32163ff91"
```
Why we are not using a secret here?

kanaagawa commented on lighthouse/run_beacon_node.sh:
```diff
@@ -0,0 +1,60 @@
+    --enr-address $EXTERNAL_IP \
+    --enr-udp-port $network_port \
+    --enr-tcp-port $network_port \
+    --port $network_port \
+	--http-address 0.0.0.0 \
+    --http-port $http_port \
```
```suggestion
    --port $network_port \
    --http-address 0.0.0.0 \
    --http-port $http_port \
```

kanaagawa commented on lighthouse/run_bootnode.sh:
```diff
@@ -0,0 +1,38 @@
+exec lighthouse boot_node \
+    --testnet-dir $TESTNET_DIR \
+    --port $BOOTNODE_PORT \
+    --listen-address 0.0.0.0 \
+	--disable-packet-filter \
+    --network-dir $DATADIR/bootnode \
```
```suggestion
    --listen-address 0.0.0.0 \
    --disable-packet-filter \
    --network-dir $DATADIR/bootnode \
```

kanaagawa commented on shared/config.go:
```diff
@@ -42,9 +42,9 @@ func UpdateChainConfig(config *params.ChainConfig) error {
-	BeaconFollowerMultiAddress = "/ip4/0.0.0.0/tcp/13001"
-	ValidatorAPI               = "http://localhost:7500"
+	BeaconAPI                  = "localhost:8000"
+	BeaconMultiAddress         = "/ip4/0.0.0.0/tcp/8000"
+	BeaconFollowerAPI          = "localhost:8001"
+	BeaconFollowerMultiAddress = "/ip4/0.0.0.0/tcp/8001"
```
```suggestion
	BeaconMultiAddress         = "/ip4/0.0.0.0/tcp/9000"
	BeaconFollowerAPI          = "localhost:8001"
	BeaconFollowerMultiAddress = "/ip4/0.0.0.0/tcp/9001"
```

kanaagawa replied:
I think the `MultiAddress` should have the same port configuration we use during the `run_beacon_node.sh` [here.](https://github.com/jimmygchen/eip4844-interop/blob/lighthouse-interop/lighthouse/run_beacon_node.sh#L39)

Inphi replied:
Yeah, this was initially done for testing.
Ideally, we can also get the mutliaddress from the beacon API so we don't have to hardcode anything:
```bash
curl -s http://127.0.0.1:8000/eth/v1/node/identity | jq .data.p2p_addresses
[
  "/ip4/172.25.0.4/tcp/9000/p2p/16Uiu2HAmRWUpDua2Sqh7cb7ftYY3qvudPvxSPF676AjLDxYkMsXQ"
]
```
This way it works for multiple beacon client configurations.

0xGabi commented on shared/jwtsecret:
```diff
@@ -0,0 +1 @@
+2481d2c18d4c305195db0d02b441970246f7368608021e37e8210ad72eef882f
```
A similar concern than before. Can we use a GitHub secret or similar here?

0xGabi reviewed (changes requested):
Great work 🤝 I did a first review of the changes while I was trying to setup the repo on my local machine. I left a few question and suggestions. I also found a couple of things that seems a bit odd to me.

>  The last component that we need to run all tests is to be able to download blobs sidecars from the beacon chain. I'm running into issues doing this using the p2p Req-Res interface.

I wonder if the p2p Req-Res issue could be related with the point I raised here: https://github.com/Inphi/eip4844-interop/pull/77#discussion_r1051320328 

---
Note that I'm having and issues during LH installation:
```
#0 141.9 warning: package `blake2 v0.10.4` in Cargo.lock is yanked in registry `crates-io`, consider running without --locked
#0 142.3  Downloading crates ...
#0 142.6 warning: spurious network error (2 tries remaining): [7] Couldn't connect to server
#0 142.6 warning: spurious network error (1 tries remaining): [7] Couldn't connect to server
#0 142.6 error: failed to compile `lighthouse v3.3.0 (/lighthouse/lighthouse)`, intermediate artifacts can be found at `/lighthouse/target`
#0 142.6
#0 142.6 Caused by:
#0 142.6   failed to download from `https://crates.io/api/v1/crates/zip/0.5.13/download`
#0 142.6
#0 142.6 Caused by:
#0 142.6   [7] Couldn't connect to server
#0 142.6 make: *** [Makefile:37: install] Error 101
------
failed to solve: executor failed running [/bin/sh -c cd lighthouse && make && make install-lcli]: exit code: 2
make: *** [lighthouse-up] Error 17
```
I will share more details after further investigation.

Inphi committed:
Couple clean up fixes

jimmygchen commented:
Hey @Inphi were you able to get the `blobtx` tests to pass? I'm getting this error while running the tests, and wonder if it's related to the issues you were seeing?

```
2022/12/19 23:55:14 checking blob from beacon node
2022/12/19 23:55:14 downloading blobs...
2022/12/19 23:55:14 getMultiAddr: failed to dial 16Uiu2HAmSifdT5QutTsaET8xqjWAMPp4obrQv7LN79f2RMmBe3nY:
  * [/ip4/0.0.0.0/tcp/8000] failed to negotiate security protocol: message did not have trailing newline
exit status 1
```

jimmygchen committed:
Fix whitespace formatting

Co-authored-by: Gabi <gabrielpk.18@gmail.com>

jimmygchen committed:
Fix whitespace formatting

Co-authored-by: Gabi <gabrielpk.18@gmail.com>

Inphi commented:
> Hey @Inphi were you able to get the `blobtx` tests to pass? I'm getting this error while running the tests, and wonder if it's related to the issues you were seeing?
> 
> ```
> 2022/12/19 23:55:14 checking blob from beacon node
> 2022/12/19 23:55:14 downloading blobs...
> 2022/12/19 23:55:14 getMultiAddr: failed to dial 16Uiu2HAmSifdT5QutTsaET8xqjWAMPp4obrQv7LN79f2RMmBe3nY:
>   * [/ip4/0.0.0.0/tcp/8000] failed to negotiate security protocol: message did not have trailing newline
> exit status 1
> ```
> 
> 

I've made some changes to https://github.com/jimmygchen/eip4844-interop/pull/2 that should fix that specific error and a couple of other issues. Please pull those changes.
However, I'm still seeing other issues retrieving blobs from req-res RPC. xposting from the discord thread:
```
issue I'm running into using `blobs_by_range`  RPC:
 Fetching the blobs from a slot containing kzg commitments fails 

Dec 16 20:30:27.425 DEBG Received BlobsByRange Request           start_slot: 77, count: 1, peer_id: 12D3KooW9pSpfniaTz44UB655hxdPX8f7BWdyPCLWee36uNA827B
Dec 16 20:30:27.425 ERRO Blob in the chain is not in the store   request_root: 0x3eec1319f0c90c96f93f8246e83386e49d9db377d3d8f868689b584511f875d2
Dec 16 20:30:27.425 DEBG BlobsByRange Response processed         returned: 0, requested: 1, current_slot: 151, start_slot: 77, msg: Failed to return all requested blocks, peer: 12D3KooW9pSpfniaTz44UB655hxdPX8f7BWdyPCLWee36uNA827B

My  (super) light client doesn't receive any data in response to the RPC, not even a p2p status code.
```

I think this is a lighthouse issue as I'm now able to download blobs using the same script from prysm. @realbigsean is taking a look

jimmygchen commented:
> I'm getting a "protocol not supported" error when attempting to create a stream for a `blobs_sidecars_by_range` request. According to the logs, the lighthouse beacon nodes do not support the `blobs_sidecars_by_range` Req-Res, or at least they aren't advertising it:
> 
> ```
> Dec 16 07:15:30.003 DEBG Identified Peer                         protocols: ["/meshsub/1.1.0", "/meshsub/1.0.0", "/eth2/beacon_chain/req/status/1/ssz_snappy", "/eth2/beacon_chain/req/goodbye/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/1/ssz_snappy", "/eth2/beacon_chain/req/ping/1/ssz_snappy", "/eth2/beacon_chain/req/metadata/2/ssz_snappy", "/eth2/beacon_chain/req/metadata/1/ssz_snappy", "/ipfs/id/1.0.0", "/ipfs/id/push/1.0.0"], observed_address: "/ip4/127.0.0.1/tcp/9001/p2p/16Uiu2HAmCHMi7ZK5T4pLepymn7fp3W4bydedBCS5Tva131EAS6AQ", listening_addresses: ["/ip4/127.0.0.1/tcp/9002", "/ip4/172.24.53.62/tcp/9002", "/ip4/127.0.0.1/tcp/9002", "/ip4/172.17.0.1/tcp/9002"], agent_version: Lighthouse/v3.3.0-1644978+/x86_64-linux, protocol_version: eth2/1.0.0, peer: 16Uiu2HAkzNWuYNSoJLBxrpT1eBRsyMRJWjhW1AworbybRqSN94Nd, service: libp2p
> ```
> 
> `/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy` is on that list, but not `/eth2/beacon_chain/req/blobs_sidecars_by_range/1/ssz_snappy` (and neither is `/eth2/beacon_chain/req/beacon_block_and_blobs_sidecar_by_root/1` which is needed for historical sync)

Hey @Inphi I think you're right, the two protocols doesn't seem to be advertised, I've attempted to fix this here @realbigsean. I'm going to try test this out.

https://github.com/sigp/lighthouse/pull/3819

jimmygchen committed:
Merge remote-tracking branch 'origin/lh-local-testnet' into lighthouse-interop

jimmygchen commented:
^ Ah, that was fixed 😅  My local branch was outdated

jimmygchen committed:
Update lighthouse version

jimmygchen committed:
Remove unused volumes

Co-authored-by: Gabi <gabrielpk.18@gmail.com>

jimmygchen committed:
Remove unused volumes

Co-authored-by: Gabi <gabrielpk.18@gmail.com>

jimmygchen committed:
read context bytes from stream

jimmygchen committed:
update geth version and enable blob tx tests for lighthouse

jimmygchen committed:
Increase build timeout

jimmygchen committed:
Increase target peer count to allow connection from tests

jimmygchen committed:
Add @Inphi's fix on download.go

jimmygchen committed:
Add missing changes

jimmygchen committed:
Test port change on CI

jimmygchen commented:
@Inphi @realbigsean 
I think the inconsistent transport errors (stream reset, EOF) I was seeing is likely related to how Lighthouse disconnects clients that doesn't support `metadata` protocol. 

Sometimes the RPC request makes it before it gets disconnected, sometimes it doesn't.

```
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
EKANS%
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
2022/12/23 04:37:58 App failed: EOF: unable to send blobs RPC request
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
EKANS%
```

Lighthouse

```
Dec 22 17:37:58.805 DEBG Failed incoming connection              error: Pending connection: Transport error on connection: , from: /ip4/192.168.0.1/tcp/52280, our_addr: /ip4/192.168.0.2/tcp/9000, service: libp2p
Dec 22 17:37:58.818 DEBG Connection established                  connection: Listener, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, service: libp2p
Dec 22 17:37:58.824 DEBG RPC Error                               direction: Outgoing, score: 0, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, client: Unknown, err: Peer does not support the protocol, protocol: metadata, service: libp2p
Dec 22 17:37:58.824 DEBG Peer score adjusted                     score: -10.00, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, msg: handle_rpc_error, service: libp2p
Dec 22 17:37:58.824 DEBG Peer does not support gossipsub         peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, service: libp2p
Dec 22 17:37:58.824 DEBG Peer transitioned to disconnect state   past_state: Healthy, score: -20.00, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, service: libp2p
Dec 22 17:37:58.824 DEBG Peer Manager disconnecting peer         reason: Bad Score, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, service: libp2p
Dec 22 17:37:58.827 DEBG Ignoring rpc message of disconnecting peer, peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, msg_kind: outbound_err, protocol: goodbye, service: libp2p
Dec 22 17:37:58.827 DEBG Peer disconnected                       peer_id: 12D3KooWGkCqEQK4Ew36LJJtULfmLHU4wnh5xbk3BBYwJTMKZQQB, service: libp2p

```

Inphi commented:
Yeah I think you're correct on the RPC client not supporting essential protocols. I'll fix the download script to do this.

jimmygchen commented:
**UPDATE**:  it turned out that the test was sending request to the primary node where it was supposed to send to the secondary, I've pushed a fix. However I wonder why the original 2nd request to primary node returns 0 sidecars.. 🤔 

~~The other issue I had was the 2nd follower node seems to take a long time to receive the blob. I think I waited around 10 seconds or more.~~

```
2022/12/23 06:02:07 checking blob from beacon node
2022/12/23 06:02:07 downloading blobs...
2022/12/23 06:02:07 checking blob from beacon node follower
2022/12/23 06:03:07 downloading blobs...
2022/12/23 06:03:08 No blobs found in requested slots, sidecar count: 0
```

Attempts on downloading blobs from node1 (9000) and node2 (9001), eventually both returned the blob.. (and so many different error messages for the same issue!)

```
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9001 --slot 65
2022/12/23 05:08:53 App failed: no blobs found in requested slots, sidecar count: 0
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9001 --slot 65
2022/12/23 05:08:59 App failed: no blobs found in requested slots, sidecar count: 0
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9001 --slot 65
2022/12/23 05:09:04 App failed: no blobs found in requested slots, sidecar count: 0
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
EKANS%
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
2022/12/23 05:09:17 App failed: EOF: unable to send blobs RPC request
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
2022/12/23 05:09:23 App failed: stream reset: unable to send blobs RPC request
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
2022/12/23 05:09:28 App failed: stream reset: readChunkedBlobsSidecar: unable to send blobs RPC request
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9000 --slot 65
EKANS%
➜  blob-utils git:(master) ✗ OA
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9001 --slot 65
EKANS%
➜  blob-utils git:(master) ✗ ./blob-utils download --beacon-p2p-addr /ip4/127.0.0.1/tcp/9001 --slot 65
EKANS%
```

jimmygchen committed:
Add temporay workaround to retry downloads, and wait longer for blob to propagate

jimmygchen committed:
Fix incorrect multiaddress used for peer

jimmygchen committed:
fix typo

jimmygchen commented:
@Inphi I've finally managed to get the `blobtx` tests passing locally after a few fixes and with the retry workaround 🥳  

```
2022/12/23 06:38:37 checking blob from beacon node
2022/12/23 06:38:37 downloading blobs...
2022/12/23 06:38:37 checking blob from beacon node follower
2022/12/23 06:38:43 downloading blobs...
2022/12/23 06:38:43 1 of 10 attempts to send blobs p2p request failed: stream reset
➜  eip4844-interop git:(lighthouse-interop) ✗ 
```

^ this is passing with retry, the 2nd attempted worked and the test completed with no non-zero exit code.

waiting for CI result now 🤞 

ps: it's still unclear to me why the node returns 0 sidecars sometimes on the 2nd attempt.. given that the 1st attempt returned one

Inphi committed:
fix download for lighthouse nodes

Inphi committed:
update fee-market test

Inphi reviewed (approved):
Superb work. Thank you!!
The blobtx test passes for me locally. And fee-market test too (with a small [patch](https://github.com/jimmygchen/eip4844-interop/pull/2/commits/46a85d32fd473917305c73471823208a5f677c1a))

jimmygchen committed:
Enable fee market tests

jimmygchen commented:
> I think it's good adding the trusted setup file to the shared directory, but would be better to use the format in the spec repo here, because I think other clients are probably using this format: https://raw.githubusercontent.com/ethereum/consensus-specs/dev/presets/mainnet/trusted_setups/testing_trusted_setups.json
> 
> Lighthouse can't actually read the JSON format at the moment though, pawan is working on making it so that we can. But until then we can point `--trusted-setup-file` to `lighthouse/scripts/local_testnet/trusted_setup.txt`

Just noticed that Lodestar is using a trusted setup file in the same format too.. lets see what the other clients do
